amt_test.go

```diff
@@ -726,6 +815,60 @@ func TestForEach(t *testing.T) {
 	if x != len(indexes) {
 		t.Fatal("didnt see enough values")
 	}
+
+	err = na.ForEach(ctx, func(i uint64, v *cbg.Deferred) error {
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, "bafy2bzaceanqxtbsuyhqgxubiq6vshtbhktmzp2if4g6kxzttxmzkdxmtipcm", c.String())
+	assert.Equal(t, bsStats{r: 1431, w: 1431, br: 88649, bw: 88649}, trackingBs.stats)
+}
+
+// Test is for specs-actors functionality that needs to be matched
+func TestForEachMutate(t *testing.T) {
+	trackingBs := newMockBlocks()
+	bs := cbor.NewCborStore(trackingBs)
+	ctx := context.Background()
+	a, err := NewAMT(bs)
+	require.NoError(t, err)
+
+	indexes := []uint64{1, 9, 66, 74, 82, 515}
+
+	for _, i := range indexes {
+		if err := a.Set(ctx, i, cborstr("value")); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	c, err := a.Flush(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	na, err := LoadAMT(ctx, bs, c)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := na.ForEach(ctx, func(i uint64, v *cbg.Deferred) error {
+		// if err := out.UnmarshalCBOR(bytes.NewReader(val.Raw)); err != nil {
+		// 	return err
+		// }
+		if i == 1 || i == 74 {
+			if err := na.Set(ctx, i, v); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, "bafy2bzaced44wtasbcukqjqicvxzcyn5up6sorr5khzbdkl6zjeo736f377ew", c.String())
+	assert.Equal(t, bsStats{r: 12, w: 12, br: 573, bw: 573}, trackingBs.stats)
```

Comment:
these stats 😱 are why I think we need to work on a collapsable form https://github.com/filecoin-project/go-amt-ipld/issues/17#issuecomment-668568876 .. I don't have any idea how common this kind of perverse sparseness is in actual Filecoin data though

Reply:
the only stat that isn't used in Filecoin was bytes read, all other 3 are relevant